### PR TITLE
Scaling integration tests

### PR DESCRIPTION
**Pull request by marceloneppel**
# Issue
Scaling up/down and also the replication of data after thoses process should be tested.

# Solution
* Add integration tests for scaling up/down and test that the data is preserved between those operations.

# Context
* Helper methods used by the methods implemented on this PR were created on https://github.com/canonical/postgresql-operator/pull/10.
* Scaling up/down logic was added on https://github.com/canonical/postgresql-operator/pull/11.
* Some test helper methods were created in this PR to organize better the test code and avoid duplicating code.

# Testing
* The new integration tests assert that the data is still being replicated after scaling down and up again.

# Release Notes
* Add tests ensuring scaling and replication work correctly.

### CODE REVIEW

File: tests/integration/test_charm.py [hunk 1]
```diff
@@ -39,59 +49,59 @@ async def test_deploy(ops_test: OpsTest, charm: str, series: str):
...
         f"dbname='postgres' user='postgres' host='{host}' password='{password}' connect_timeout=1"
-    ) as connection, connection.cursor() as cursor:
+    ) as connection:
         assert connection.status == psycopg2.extensions.STATUS_READY
 
         # Retrieve settings from PostgreSQL pg_settings table.
         # Here the SQL query gets a key-value pair composed by the name of the setting
         # and its value, filtering the retrieved data to return only the settings
         # that were set by Patroni.
+        cursor = connection.cursor()
```
Comment (WRFitch) [inline tests/integration/test_charm.py:104]: This could probably be closed after line 111

**Review by WRFitch — APPROVED** [review]
LGTM!

File: tests/integration/test_charm.py [hunk 2]
```diff
@@ -118,15 +128,60 @@ async def test_settings_are_correct(ops_test: OpsTest, series: str):
...
+    await ops_test.model.applications[application_name].add_unit(count=1)
+    await ops_test.model.wait_for_idle(apps=[application_name], status="active", timeout=1000)
+
+    # Testing write occurred to every postgres instance by reading from them
+    for unit in ops_test.model.applications[application_name].units:
+        host = unit.public_address
+        logger.info("connecting to the database host: %s", host)
+        with db_connect(host=host, password=password) as connection:
+            # Ensure we can read from "leaderdeletiontest" table
+            connection.cursor().execute("SELECT * FROM leaderdeletiontest;")
```
Comment (MiaAltieri) [inline tests/integration/test_charm.py:187]: add checks for when the primary gets removed that the configuration is accurate (ie the juju unit ips match the PG ips)

File: tests/integration/test_charm.py [hunk 3]
```diff
@@ -118,15 +128,60 @@ async def test_settings_are_correct(ops_test: OpsTest, series: str):
...
 
-def convert_records_to_dict(records: List[tuple]) -> dict:
-    """Converts pyscopg2 records list to a dict."""
-    dict = {}
-    for record in records:
-        # Add record tuple data to dict.
-        dict[record[0]] = record[1]
-    return dict
+@pytest.mark.parametrize("series", SERIES)
+async def test_persist_data_through_leader_deletion(ops_test: OpsTest, series: str):
```
Comment (MiaAltieri) [inline tests/integration/test_charm.py:156]: nit change this to `test_persist_data_through_primary_deletion`

File: tests/integration/test_charm.py [hunk 4]
```diff
@@ -118,15 +128,60 @@ async def test_settings_are_correct(ops_test: OpsTest, series: str):
...
+    """Test data is replicated to new units after a scale up."""
+    # Set a composite application name in order to test in more than one series at the same time.
+    application_name = build_application_name(series)
+
+    # Ensure the initial number of units in the application.
+    initial_scale = len(UNIT_IDS)
+    await scale_application(ops_test, application_name, initial_scale)
+
+    # Scale down the application.
+    await scale_application(ops_test, application_name, initial_scale - 1)
```
Comment (MiaAltieri) [inline tests/integration/test_charm.py:142]: if possible include the leader unit and the primary unit in this scale down, then you can ignore my comments

-  "add checks for when the primary gets removed that the configuration is accurate (ie the juju unit ips match the PG ips)"
- "add a test for removing the juju leader"

File: tests/integration/test_charm.py (hunk 2 above)
Comment (MiaAltieri) [inline tests/integration/test_charm.py:187]: add a test for removing the juju leader (ie the juju unit ips match the PG ips)

**Review by MiaAltieri — CHANGES_REQUESTED** [review]

File: tests/integration/test_charm.py (hunk 1 above)
Comment (shayancanonical) [reply tests/integration/test_charm.py:104]: I wonder if cursors are closed automagically when the connection is closed (when exiting the context manager)

**Review by shayancanonical — APPROVED** [review]

**Review by paulomach — APPROVED** [review]
lgtm, not a lot to add

File: tests/integration/test_charm.py (hunk 1 above)
Comment (marceloneppel) [reply tests/integration/test_charm.py]: Good catch! Cursors are closed when exiting the context manager, but I figured out that connection dont. SoI fixed it using the syntax explained in https://github.com/psycopg/psycopg2/blob/f469331af5f7b236f8037b302b69c2bf0df2a096/doc/src/connection.rst (keeping the connection in different context managers to have scope for transactions, and closing the connection later).

File: tests/integration/test_charm.py (hunk 3 above)
Comment (marceloneppel) [reply tests/integration/test_charm.py]: Fixed on https://github.com/canonical/postgresql-operator/pull/14/commits/3cadd303971eda2ab85ffa8e57840b176d66e9fb.

File: tests/integration/test_charm.py (hunk 2 above)
Comment (marceloneppel) [reply tests/integration/test_charm.py]: Added on https://github.com/canonical/postgresql-operator/pull/14/commits/6fe523270b2b5a4424f9ed91c7038de88e66812c (and also on https://github.com/canonical/postgresql-operator/pull/14/commits/b23967277d37d7d2291c48fe0653f71954b09f91, that contains additional fixes).

File: tests/integration/test_charm.py (hunk 4 above)
Comment (marceloneppel) [reply tests/integration/test_charm.py:145]: Added on https://github.com/canonical/postgresql-operator/pull/14/commits/6fe523270b2b5a4424f9ed91c7038de88e66812c (and also on https://github.com/canonical/postgresql-operator/pull/14/commits/b23967277d37d7d2291c48fe0653f71954b09f91, that contains additional fixes).

File: tests/integration/test_charm.py (hunk 2 above)
Comment (marceloneppel) [reply tests/integration/test_charm.py]: Added on https://github.com/canonical/postgresql-operator/pull/14/commits/6fe523270b2b5a4424f9ed91c7038de88e66812c (and also on https://github.com/canonical/postgresql-operator/pull/14/commits/b23967277d37d7d2291c48fe0653f71954b09f91, that contains additional fixes).

**Review by MiaAltieri — APPROVED** [review]

File: tests/integration/test_charm.py [hunk 5]
```diff
@@ -39,67 +52,67 @@ async def test_deploy(ops_test: OpsTest, charm: str, series: str):
...
 
     # Connect to PostgreSQL.
-    host = ops_test.model.units.get(f"{application_name}/0").public_address
+    host = get_unit_address(ops_test, f"{application_name}/{unit_id}")
     logger.info("connecting to the database host: %s", host)
-    with psycopg2.connect(
-        f"dbname='postgres' user='postgres' host='{host}' password='{password}' connect_timeout=1"
-    ) as connection, connection.cursor() as cursor:
+    connection = db_connect(host, password)
+    with connection:
```
Comment (WRFitch) [inline tests/integration/test_charm.py:99]:
```suggestion
with db_connect(host, password) as connection:
```

File: tests/integration/test_charm.py [hunk 6]
```diff
@@ -118,15 +131,127 @@ async def test_settings_are_correct(ops_test: OpsTest, series: str):
...
+    application_name = build_application_name(series)
+    any_unit_name = ops_test.model.applications[application_name].units[0].name
+    primary = await get_primary(ops_test, any_unit_name)
+    password = await get_postgres_password(ops_test, primary)
+
+    # Write data to primary IP.
+    host = get_unit_address(ops_test, primary)
+    logger.info(f"connecting to primary {primary} on {host}")
+    connection = db_connect(host, password)
+    with connection:
```
Comment (WRFitch) [inline tests/integration/test_charm.py:232]:
```suggestion
with db_connect(host, password) as connection:
```

File: tests/integration/test_charm.py [hunk 7]
```diff
@@ -118,15 +131,127 @@ async def test_settings_are_correct(ops_test: OpsTest, series: str):
...
-    for record in records:
-        # Add record tuple data to dict.
-        dict[record[0]] = record[1]
-    return dict
+    # Testing write occurred to every postgres instance by reading from them
+    for unit in ops_test.model.applications[application_name].units:
+        host = unit.public_address
+        logger.info("connecting to the database host: %s", host)
+        connection = db_connect(host, password)
+        with connection:
```
Comment (WRFitch) [inline tests/integration/test_charm.py:253]:
```suggestion
with db_connect(host, password) as connection:
```

**Review by WRFitch — APPROVED** [review]
LGTM, minor style comment about context managers but feel free to ignore it.

File: tests/integration/test_charm.py (hunk 5 above)
Comment (marceloneppel) [reply tests/integration/test_charm.py]: Fixed on https://github.com/canonical/postgresql-operator/pull/14/commits/08a1aa1d025239f23c039b80fcdf41b4bfe83cad.

File: tests/integration/test_charm.py (hunk 6 above)
Comment (marceloneppel) [reply tests/integration/test_charm.py]: Fixed on https://github.com/canonical/postgresql-operator/pull/14/commits/08a1aa1d025239f23c039b80fcdf41b4bfe83cad.

File: tests/integration/test_charm.py (hunk 7 above)
Comment (marceloneppel) [reply tests/integration/test_charm.py]: Fixed on https://github.com/canonical/postgresql-operator/pull/14/commits/08a1aa1d025239f23c039b80fcdf41b4bfe83cad.

**Pull request merged**